agent/service.go

```diff
@@ -136,44 +218,33 @@ func proxyIO(ctx context.Context, path string, port uint32, in bool) {
 		}
 		break
 	}
-	go func() {
-		<-ctx.Done()
-		conn.Close()
-		f.Close()
-	}()
-	log.G(ctx).Debug("begin copying io")
+
 	buf := make([]byte, internal.DefaultBufferSize)
-	if in {
-		_, err = io.CopyBuffer(f, conn, buf)
+
+	// TODO useless log without context
+	log.G(ctx).Debug("begin copying io")
+
+	if isStdin {
+		_, err = io.CopyBuffer(fifo, conn, buf)
 	} else {
-		_, err = io.CopyBuffer(conn, f, buf)
-	}
-	if err != nil {
-		log.G(ctx).WithError(err).Error("error with stdio")
+		_, err = io.CopyBuffer(conn, fifo, buf)
 	}
-}
 
-func unpackBundle(path string, bundle *types.Any) (*types.Any, error) {
-	// get json bytes from task request
-	extraData := &proto.ExtraData{}
-	err := types.UnmarshalAny(bundle, extraData)
 	if err != nil {
-		return nil, err
-	}
-	// write bundle/config.json bytes
-	err = ioutil.WriteFile(path, extraData.JsonSpec, 0644)
-	if err != nil {
-		return nil, err
+		log.G(ctx).WithError(err).Error("error with stdio")
 	}
-	return extraData.RuncOptions, nil
 }
 
 // State returns process state information
 func (ts *TaskService) State(ctx context.Context, req *shimapi.StateRequest) (*shimapi.StateResponse, error) {
 	log.G(ctx).WithFields(logrus.Fields{"id": req.ID, "exec_id": req.ExecID}).Debug("state")
+	runcService, ok := ts.containers[req.ID]
```

Comment:
Should this be wrapped with `containersMu` as well? (or even `RWMutex`).

Reply:
Oh good catch, yeah I'll probably just make a getContainer() method that gets the lock (and uses RWMutex)